packages/bull-vault/src/BullStrategy.sol

```diff
@@ -157,10 +176,56 @@ contract BullStrategy is ERC20, LeverageBull {
         emit Withdraw(msg.sender, _bullAmount, wPowerPerpToRedeem);
     }
 
+    /**
+     * @notice redeem the Crab shares owned by Bull if Squeeth contracts are shutdown and collapse leverage trade to hold ETH only
+     * @param _params Shutdown params
+     */
+
+    function redeemShortShutdown(ShutdownParams calldata _params) external onlyOwner {
+        hasRedeemedInShutdown=true;
+        ICrabStrategyV2(crab).withdrawShutdown(ICrabStrategyV2(crab).balanceOf(address(this)));
+        uint256 usdcToRepay = _calcUsdcToRepay(ONE);
+        _exactOutFlashSwap(
+            weth,
+            usdc,
+            _params.ethPoolFee,
+            usdcToRepay,
+            _params.maxEthToPay,
+            uint8(FLASH_SOURCE.SHUTDOWN),
+            abi.encodePacked("")
+        );
+    }
+    /**
+     * @notice allows a user to withdraw their share of ETH if squeeth contracts have been shutdown
+     * @dev redeemShortShutdown must have been called first
+     * @param _bullAmount bull amount to withdraw
+     */
+    function withdrawShutdown(uint256 _bullAmount) external {
+        require (hasRedeemedInShutdown, "BS3");
+        uint256 share = _bullAmount.wdiv(totalSupply());
+        uint256 ethToReceive = share.wmul(address(this).balance);
+        _burn(msg.sender, _bullAmount);
+        payable(msg.sender).sendValue(ethToReceive);
+    }
+
```

Comment:
Is it possible to move this to another separate module?

like having BullStrategy have an `emergencyWithdraw` function that withdraw funds, can only be called by `EmergencyShutdown` contract.

and then the Emergency shutdown contract implement everything else? I just think it is better from design perspective to keep the BullStrategy simple without the flash swap structs and stuff...

Wyt?

Reply:
yeah, let me look at that. i did struggle a bit with the dependencies for this as it is the only part of BullStrategy that needs uni flashswap stuff